Merge of haskell-with-nixpkgs back into HNix

Anton-Latukha opened the pull request:
This PR update build processes, both with CI processes:
  *  Migrates project to the GitHub Actions CI
    * :heavy_check_mark: updates from Travis 50min restriction to GitHub 6 hours of execution time, which allows churning through Nixpkgs revisions that when Hydra CI Cache is not present (since Hydra CI cache has a pretty constant lag to after the channel update).
    * :heavy_check_mark: faster node startup times
    * :heavy_check_mark: much bigger number of nodes and the compute power of nodes
    * :heavy_check_mark: Cachix and Nix deployment processes are better supported for GitHub Actions.
    * :heavy_check_mark: multiple checks reported to PR, so the community can roughly see the info on build results even without going into CI interface.  #593
  * :heavy_check_mark: expands the number of builds: https://github.com/haskell-nix/hnix/pull/681/checks
  * :heavy_check_mark: allows to better customize and tune the build processes
  * :heavy_check_mark: introduces classical Cabal Hackage builds (`8.{4,10}`), that is mandatory for the Haskell project, #653
  * CI updates:
    * :heavy_check_mark: Nixpkgs GHC 8.8.(3->4), #683 
    * :heavy_check_mark: Nixpkgs drop of GHC 8.4.4, migration 8.4.4 to Hackage build, #684
  * etc...

Also:
  *  :heavy_check_mark:  While waiting on GitHub CI Cachix key setup (#682), reduced Travis, all builds become optional and added "bypass build" for fast skip, it is currently ours "green" background Cachix builder.

---

`HNix` was taken and with maximal preservation of Git history, Git history was
passed through pass filter and that in how `haskell-with-nix` was formed.

Further, work on CI and build configs happened there.

Due to `HNix` being a mothership project and since `haskell-with-nix`
preservation of original history - it is possible to merge progress from it
back.

Now, when `haskell-with-nixpkgs` main setup stabilized - it is ready to be
merged back, meanwhile `HNix` needed to dodge Nixpkgs downstream quirks which
blocked the development processes and new CI was needed.

Merge went pretty ideal, for merge `haskell-with-nixpkgs` only relevant history
was also filtered-out and also dropped 3 commits that got into the passed files.

A custom merge commit also was required.

But overall, the process of merge went unexpectedly great.

---

After the inside PR merge - also the change of stub names used in
`haskell-with-nixpkgs` to `HNix`-specific names is obviously needed. So that directly follows this
merge inside PR branch.

Anton-Latukha committed:
CI: Travis: env: global: rev: upd: note for haskell-update branch

M  .travis.yml

Anton-Latukha committed:
CI: Travis: build.sh: BUILD_PROJECT: *Optparse*Completion: upd: directional note

M  build.sh

sjakobi committed:
Travis: Add CI job with linux and GHC-8.4.4 (#603)

GHC 8.4 is still widely used, so it's good to check compatibility
with it.

Anton-Latukha committed:
default.nix: make GHC 8.8.3 default

M  default.nix

Anton-Latukha committed:
default.nix: rm allowUnfree - there is none

M  default.nix

Anton-Latukha committed:
default.nix: add haskell.lib profiling options explicitly

Integrates with CI.

M  default.nix

Anton-Latukha committed:
default.nix: add flags {doCheck,doHaddock,doCoverage}

M  default.nix

Anton-Latukha committed:
Fix CI builds on internal branches (#604)

* CI: Travis: rm cache | Builds 2X times faster without Travis cache

M  .travis.yml

* CI: Travis: build.sh: mv `set` declaration | fx master CI statuses

M  .travis.yml
M  build.sh

Anton-Latukha committed:
CI: Travis: build.sh: init CACHIX_SIGNING_KEY with default empty value

Due to new strict `set` rules `if` statement check agains empty variable was
failing the build, this fixes builds in PRs.

...sigh.

M  build.sh

Anton-Latukha committed:
default.nix: expose more features of `haskell.lib` API

M  default.nix

Anton-Latukha committed:
default.nix: add docs to the exposed haskell.lib options

M  default.nix

Anton-Latukha committed:
CI: build.sh: rm empty args from nix-build

M  build.sh

Anton-Latukha committed:
CI: build.sh: update Nix after its installation

Nix that installs through Nix installer is 2.0.4.
Current Nix release is 2.3.5.

To to have a less of old bugs, and more of new bugs - Nix after installation
needs an update right away.

Old Nix literally prevented importing Nix channels on Linux Nix installations.

M  build.sh

Anton-Latukha committed:
default.nix: to allow use of current channels remove hash guarding

M  default.nix

Anton-Latukha committed:
default.nix: set `rev` default to `nixpkgs-unstable`

M  default.nix

Anton-Latukha committed:
default.nix: for import form separate explicit override set

M  default.nix

Anton-Latukha committed:
default.nix: add if useRev - then use rev, else use <nixpkgs>

M  default.nix

Anton-Latukha committed:
CI: build.sh: MAIN: add the channel update

M  build.sh

Anton-Latukha committed:
default.nix: expose additional part of Nixpkgs haskell.lib API

M  default.nix

Anton-Latukha committed:
default.nix: default enableDeadCodeElimination to false | GHC macOS build bug

M  default.nix

Anton-Latukha committed:
default.nix: off enableSharedExecutables | rare feature, not passes macOS tests

This feature seems to enable binaries as also a dynamic library, with is a rare
and largely unknown feature, and it not used. Since it is unexpected optional
rare not standard feature - turn it off by default.

M  default.nix

Anton-Latukha committed:
default.nix: m doc upd

M  default.nix

Anton-Latukha committed:
default.nix: upd docs for Header.lib args in headers, sort

Providing docs to them and sort them roughly in order of their application/stage

M  default.nix

Anton-Latukha committed:
default.nix: package: sort inherited Haskell.lib args, provide doc

M  default.nix

Anton-Latukha committed:
default.nix: add option enableSeparateBinOutput

M  default.nix

Anton-Latukha committed:
default.nix: abstract the developPackage into value

M  default.nix

Anton-Latukha committed:
default.nix: allow apply composition over the package

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: sdistTarball; add doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: buildFromSdist; add doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: failOnAllWarnings; add doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: buildStrictly; add doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: disableOptimization; add doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: linkWithGold; add doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: enableDWARFDebugging; add doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: doJailbreak

Argument was already exposed in the header

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: doStrip

Argument and doc was already provided in the header

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: justStaticExecutables

Argument and doc was already provided in the header.

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: checkUnusedPackages; provide doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: gen*Optparse*Completion; provide doc

M  default.nix

Anton-Latukha committed:
default.nix: expose functionality: doHyperlinkSource; provide doc

M  default.nix

Anton-Latukha committed:
default.nix: rm doStrict (superceeded by failOnAllWarnings)

M  default.nix

Anton-Latukha committed:
default.nix: rm doOptimize (superceeded by disableOptimizations)

M  default.nix

Anton-Latukha committed:
default.nix: doc: add beginning doc and link to lib for more info on options

M  default.nix

Anton-Latukha committed:
default.nix: allow apply composition over the package

M  default.nix

Anton-Latukha committed:
default.nix: doc: add to body (functions fold and composePackage)

M  default.nix

Anton-Latukha committed:
default.nix: doc: add to header args

M  default.nix

Anton-Latukha committed:
CI: Travis: add Nix packages upd

M  .travis.yml

Anton-Latukha committed:
CI: Travis: fix Cachix nix-env install rights menegement

The Cachix suddenly started to fall with:

```
$ nix-env -iA cachix -f https://cachix.org/api/v1/install

error: creating directory '/Users/travis/.cache/nix/tarballs': Permission denied

The command "nix-env -iA cachix -f https://cachix.org/api/v1/install" exited with 1.
```

There are changes to macOS procedures, Nix updated to 2.2, and Cachix has big
upstream changes for macOS processes.

So fixing rights management in this way.

M  .travis.yml

Anton-Latukha committed:
CI: Travis: rm sleep 2 - redundant

Anyway it takes 30s-1m to the nix-build.

M  .travis.yml

Anton-Latukha committed:
CI: Travis: add HACK to fix reload Nix-daemon after Nix update 2.0 -> 2.2

M  .travis.yml

Anton-Latukha committed:
CI: Travis: explicitly state the builds

M  .travis.yml

Anton-Latukha committed:
CI: Travis: env: rm jobs matrix

M  .travis.yml

Anton-Latukha committed:
CI: Travis: rm `os:` and `dist:`, since they are explicitly put below

M  .travis.yml

Anton-Latukha committed:
CI: Travis: enable strict build in GHC 8.6.5

M  .travis.yml

Anton-Latukha committed:
CI: Travis: upd explicit build names

M  .travis.yml

Anton-Latukha committed:
CI: Travis: make GHC 8.10 mandatory

M  .travis.yml

Anton-Latukha committed:
CI: Travis: move SDist check to GHC 8.10

M  .travis.yml

Anton-Latukha committed:
CI: .travis.tml: clean-up

Since conf no longer multiplies rows into matrix, but provides it explitly,
exclusions not needed.

M  .travis.yml

Anton-Latukha committed:
default.nix: clean-up docs

M  default.nix

Anton-Latukha committed:
build.sh: default to GHC 8.10

M  build.sh

Anton-Latukha committed:
build.sh: sort the options into logical order

M  build.sh

Anton-Latukha committed:
build.sh: provide docs to options

M  build.sh

Anton-Latukha committed:
build.sh: add disclamer for options

M  build.sh

Anton-Latukha committed:
build.sh: add more options and their docs

M  build.sh

Anton-Latukha committed:
CI: .travis.yml: use linkWithGold in Linux builds, a bit faster builds

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: GHC 8.10: add Haddock, optimize, Benchmark, Shell Complettions

M  .travis.yml

Anton-Latukha committed:
CI: make generate*Complettions work and accept a list of executables

M  .travis.yml
M  build.sh
M  default.nix

Anton-Latukha committed:
CI: rm buildStackProject, it does not work that way, and better to use real stack

M  .travis.yml
M  build.sh

Anton-Latukha committed:
build.sh: add new options into the build commands

M  build.sh

Anton-Latukha committed:
build.sh: rm `pkgName` (not used), return `name` - used by Cachix

M  build.sh

Anton-Latukha committed:
default.nix: clean-up no longer relevant commented-out code

buildInputs populated automatically.

Environment overloads we can do any time, no need to keep trivial code
commented.

There is no AWS S3 use. If we would need we would write new deployment
implementation  outside of `default.nix` using actual NixOps.

sjakobi committed:
Silence redundant import warnings for Control.Monad.Fail (#633)

Also address mysterious redundant import warnings in the testsuite

This allows us to enable `buildStrictly=true` for GHC 8.8.

Fixes #613.

Anton-Latukha committed:
CI: .travis.yml: rm Strict from GHC 8.6.5

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: properly name GHC 8.8.3 macOS as Strict

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: add explicit quotes

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: add description to the channels

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: add description for executableNamesToShellComplete

M  .travis.yml

sorki committed:
default.nix: filter package root with nix-gitignore

so we don't copy stuff like `dist-newstyle`.

Anton-Latukha committed:
CI: .travis.yml: add the disclaimer

M  .travis.yml

Anton-Latukha committed:
CI: mv rev doc (.travis.yml->build.sh)

M  .travis.yml
M  build.sh

Anton-Latukha committed:
CI: build.sh: upd disclaimer

M  build.sh

Anton-Latukha committed:
CI: build.sh: add setting section disclaimer

M  build.sh

Anton-Latukha committed:
CI: build.sh: upd NIX_PATH doc

M  .travis.yml
M  build.sh

Anton-Latukha committed:
CI: .travis.yml: reduce not used pkgName

getOptParseApplicativeComplettions does not need it

M  .travis.yml

Anton-Latukha committed:
CI: {.travis.yml,build.sh}: rename (name->project) more unique and descriptive

M  .travis.yml
M  build.sh

Anton-Latukha committed:
CI: {.travis.yml,build.sh}: merge documentation into buildh.sh

M  .travis.yml
M  build.sh

Anton-Latukha committed:
CI: .travis.yml: rm NOTEs from less important explanations

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: {add,set,sort} features from `build.sh` explicitly

M  .travis.yml

Anton-Latukha committed:
CI: build.sh: mv doHyperlinkSource into logical place

M  build.sh

Anton-Latukha committed:
CI: .travis.yml: upd the disclaimer

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: rm ALL_TESTS, since it does not work for nix-build

and also just plainly does not passes itself, so it is useless in CI build setup

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: upd CACHIX_SIGNING_KEY explanation

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: GHC 8.10: add Haddock and sources hyperlinking

M  .travis.yml

Anton-Latukha committed:
CI: {.travis.yml,build.sh}: rename (GHCVERSION->compiler)

M  .travis.yml
M  build.sh

Anton-Latukha committed:
CI: default.nix: add `compiler` explanation

M  default.nix

Anton-Latukha committed:
CI: {.travis.yml,build.sh}: upd: make conf work with new Nix installer

M  .travis.yml
M  build.sh

Anton-Latukha committed:
CI: .travis.yml: rm irrelevant remount

M  .travis.yml

Anton-Latukha committed:
CI: {build.sh,default.nix}: upd settings structure, upd descriptions

M  build.sh
M  default.nix

Anton-Latukha committed:
CI: default.nix: upd header disclamer

M  build.sh

Anton-Latukha committed:
CI: mv rev doc (build.sh->default.nix)

M  build.sh
M  default.nix

Anton-Latukha committed:
CI: merge docs (build.sh->default.nix)

M  build.sh
M  default.nix

Anton-Latukha committed:
CI: build.sh: rm most of NOTE tags

M  build.sh

Anton-Latukha committed:
CI: .travis.yml: rm Nix updates

After refactoring they are right beside/before build.sh Nix updates, so this
update can be reduced.

M  .travis.yml

Anton-Latukha committed:
CI: .travis.yml: upd service restart

Since service is not touched the daemon-reload is probably not needed, restart
is sufficient.

M  .travis.yml

Anton-Latukha committed:
CI: build.sh: add report of Nixpkgs channel revision

Since Nix and Nixpkgs channel were just updated and build is next - report the
revision.

M  build.sh

Anton-Latukha committed:
default.nix: use dummy package name

Anton-Latukha committed:
default.nix: upd rephrase

Anton-Latukha committed:
default.nix: upd rephrase

M  default.nix

Anton-Latukha committed:
default.nix: add explanation

M  default.nix

Anton-Latukha committed:
default.nix: clean-up

M  default.nix

Anton-Latukha committed:
default.nix: upd

M  default.nix

Anton-Latukha committed:
default.nix: upd

M  default.nix

Anton-Latukha committed:
default.nix: add explanation

M  default.nix

Anton-Latukha committed:
add initial .github/workflows/test.yml

Anton-Latukha committed:
M  .github/workflows/test.yml

Anton-Latukha committed:
test.yml: upd name: (Test->CI)

M  .github/workflows/test.yml

Anton-Latukha committed:
test.yml: add env of Nixpkgs Haskell Lib from HNix

M  .github/workflows/test.yml

Anton-Latukha committed:
test.yml: (nix-build->./build.sh)

M  .github/workflows/test.yml

Anton-Latukha committed:
test.yml: upd update of Nix to update all Nix installed packages

M  build.sh

Anton-Latukha committed:
test.yml: jobs: tests->Basic_build

M  .github/workflows/test.yml

Anton-Latukha committed:
{test.yml,build.sh}: project: hnix->replaceWithProjectNameInCachix

M  .github/workflows/test.yml
M  build.sh

Anton-Latukha committed:
{test.yml,build.sh}: executableNames..Complete: hnix->replaceWith...

M  .github/workflows/test.yml
M  build.sh

Anton-Latukha committed:
test.yml: jobs: Basic_build: name the steps

M  .github/workflows/test.yml

Anton-Latukha committed:
build.sh: compiler: add explanation

M  build.sh

Anton-Latukha committed:
default.nix: compiler: mv explanation

M  build.sh
M  default.nix

Anton-Latukha committed:
test.yml: m upd

M  .github/workflows/test.yml

Anton-Latukha committed:
test.yml: trigger CI on push to master, track only PRs| trigger build once

M  .github/workflows/test.yml

Anton-Latukha committed:
Simplify step names

M  .github/workflows/test.yml

Anton-Latukha committed:
GitHub CI: Nixpkgs: disable withHoogle

M  .github/workflows/test.yml

Anton-Latukha committed:
GitHub CI: add Hackage_Cabal

M  .github/workflows/test.yml

Anton-Latukha committed:
.github/workflows/(test->main).yml

R  .github/workflows/test.yml -> .github/workflows/main.yml

Anton-Latukha committed:
builds: rname project->cachixAccount

M  .github/workflows/main.yml
M  build.sh

Anton-Latukha committed:
.travis.yml: add withHoogle=false for CI builds

M  .travis.yml

Anton-Latukha committed:
.travis.yml: group cachix conf

M  .travis.yml

Anton-Latukha committed:
GitHub CI: add name tags to builds

M  .github/workflows/main.yml

Anton-Latukha committed:
build.sh: provide a NIX_PATH change if rev is chosen

M  build.sh

Anton-Latukha committed:
GitHub CI: add Linux_GHC-8-10

M  .github/workflows/test.yml

Anton-Latukha committed:
GitHub CI: add Linux_GHCJS

M  .github/workflows/test.yml

Anton-Latukha committed:
GitHub CI: add macOS_default

M  .github/workflows/test.yml

Anton-Latukha committed:
GitHub CI: Hackage_Cabal: add cabal v2-build

M  .github/workflows/test.yml

Anton-Latukha committed:
GitHub CI: Hackage_Cabal: add cabal v2-test

M  .github/workflows/test.yml

Anton-Latukha committed:
GitHub CI: Nixpkgs custom build: add customization

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: Hackage_Cabal: add cabal v2-configure

M  .github/workflows/test.yml

Anton-Latukha committed:
GitHub CI: add Nixpkgs_stable_Linux

M  .github/workflows/main.yml

Anton-Latukha committed:
builds: pass-through useRev

M  .github/workflows/main.yml
M  build.sh
M  default.nix

Anton-Latukha committed:
builds: use rev

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: add build schedule

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: add explanations

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: make all builds optional

M  .github/workflows/main.yml

Anton-Latukha committed:
name Hackage_Cabal

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: Nixpkgs_unstable_macOS: upd

M  .github/workflows/main.yml

Anton-Latukha committed:
build.sh: rm NIX_PATH setup| default.nix solves it internally

M  build.sh

Anton-Latukha committed:
build.sh: short-curcuit the Nix tools update

M  build.sh

Anton-Latukha committed:
add .github/workflows/Cabal-Linux.yml

Anton-Latukha committed:
mv Hackage_Cabal build main->Cabal-Linux.yml

M  .github/workflows/Cabal-Linux.yml
M  .github/workflows/main.yml

Anton-Latukha committed:
Cabal-Linux.yml: add schedule

M  .github/workflows/Cabal-Linux.yml

Anton-Latukha committed:
main.yml: switch all Nix installer nix_path to nixos-unstable

M  .github/workflows/main.yml

Anton-Latukha committed:
main.yml: rm all Nix installer nix_path resetting to default

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: add macOS, mv job into it

A  .github/workflows/Nixpkgs-macOS.yml
M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: Nixpkgs-macOS: add triggers and shedule

M  .github/workflows/Nixpkgs-macOS.yml

Anton-Latukha committed:
GitHub CI: workflow: Nixpkgs-macOS: m build rename

M  .github/workflows/Nixpkgs-macOS.yml

Anton-Latukha committed:
GitHub CI: workflow: Nixpkgs-GHCJS: mv build in

A  .github/workflows/Nixpkgs-GHCJS.yml
M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: rm withHoogle overload since it is redundant

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: (CI->Nixpkgs, Linux)

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: linkWithGold

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd build ids

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd build names

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: add NixOS-unstable Strict build

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd build names

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: add build GHC 8.6.5

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: add build GHC 8.4.4

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd build name

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd build names, wrap in "

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: main: upd name: add GHC for GitHub CI to not mix-in GHCJS

M  .github/workflows/main.yml

Anton-Latukha committed:
GitHub CI: workflow: mv (main->Nixpkgs-GHC-Linux).yml

R  .github/workflows/main.yml -> .github/workflows/Nixpkgs-GHC-Linux.yml

Anton-Latukha committed:
GitHub CI: workflow: mv Quality build

M  .github/workflows/Nixpkgs-GHC-Linux.yml

Anton-Latukha committed:
GitHub CI: workflow: m rname builds

M  .github/workflows/Nixpkgs-GHC-Linux.yml

Anton-Latukha committed:
GitHub CI: workflow: m rename of the build

M  .github/workflows/Nixpkgs-GHC-Linux.yml

Anton-Latukha committed:
GitHub CI: workflow: Quality build: add nix-shell test

M  .github/workflows/Nixpkgs-GHC-Linux.yml

Anton-Latukha committed:
GitHub CI: workflow: add denotation towards main examples

M  .github/workflows/Nixpkgs-GHC-Linux.yml

Anton-Latukha committed:
GitHub CI: workflow: Quality build: fx nix-shell command

M  .github/workflows/Nixpkgs-GHC-Linux.yml

Anton-Latukha committed:
GitHub CI: workflow: mv Nixpkgs-(GHC-->)Linux.yml

R  .github/workflows/Nixpkgs-GHC-Linux.yml -> .github/workflows/Nixpkgs-Linux.yml

Anton-Latukha committed:
GitHub CI: workflow: GHCJS: clean-up

M  .github/workflows/Nixpkgs-GHCJS.yml

Anton-Latukha committed:
GitHub CI: workflow: GHCJS: use nixos-unstable rev

M  .github/workflows/Nixpkgs-GHCJS.yml

Anton-Latukha committed:
GitHub CI: workflow: GHCJS: use linkWithGold

M  .github/workflows/Nixpkgs-GHCJS.yml

Anton-Latukha committed:
R  .github/workflows/Nixpkgs-Linux.yml -> .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: Nixpgs, Linux: recommend nixos-unstable

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: Nixpgs, Linux: recommend test for nixos-20.03 stable

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: mv additiona builds into Nixpgs-Linux-additional

A  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: rname workflows

M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: make main builds mandatory

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: make build not strict, to not scare off people ;)

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: make Nix-shell into a separate test

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: Nix-shell: from the hip nix-shell'ing

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: Nix-shell: add separate Hoogle test

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: rename GHCJS

M  .github/workflows/Nixpkgs-GHCJS.yml

Anton-Latukha committed:
GitHub CI: workflow: main: prettify

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: fx Nix-shell test

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: rename Nix-shell build

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: m upd

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd Nix-shell build

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd Quality build

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd NixOS stable build

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd disclaimer

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: upd steps to "Determined Nix-build"

M  .github/workflows/Nixpkgs-GHCJS.yml
M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml
M  .github/workflows/Nixpkgs-macOS.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd Hoogle step name

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: m upd

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd Hoogle step name

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd Hoogle build name

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: upd Basic build name

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: add explicit nixos-unstable

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: Basic: add nixos-unstable explanation

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
GitHub CI: workflow: main: Nix-shell: use pure shell (*sic)

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
default.nix: abstract over package root| now able to manage monorepos

Anton-Latukha committed:
default.nix: abstract over package name| full support of monorepos

M  default.nix

Anton-Latukha committed:
build.sh: passthrough package{Name,Root} into default.nix

Anton-Latukha committed:
{default.nix,build.sh}: (package->cabal)Name end explanation

Anton-Latukha committed:
CI: Nixpkgs-Linux-main: add template for monorepo matrix

Anton-Latukha committed:
CI: Cabal-Linux: add template for monorepo matrix

Anton-Latukha committed:
D  ghcjs/default.nix

Anton-Latukha committed:
Nixpkgs-macOS: mv env attrset

Anton-Latukha committed:
CI: Nixpkgs workflows: use haskell-with-nixpkgs cachix account

M  .github/workflows/Nixpkgs-GHCJS.yml
M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml
M  .github/workflows/Nixpkgs-macOS.yml
M  .travis.yml

Anton-Latukha committed:
CI: Nixpkgs workflows: provide CACHIX_SIGNING_KEY

M  .github/workflows/Nixpkgs-GHCJS.yml
M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml
M  .github/workflows/Nixpkgs-macOS.yml
M  .travis.yml

Anton-Latukha committed:
CI: GitHub: Nixpkgs workflows: Cachix action, provide account to it

M  .github/workflows/Nixpkgs-GHCJS.yml
M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml
M  .github/workflows/Nixpkgs-macOS.yml

Anton-Latukha committed:
CI: GitHub: Nixpkgs-Linux-main: quality build: upd name

Anton-Latukha committed:
Merge progress in haskell-with-nixpkgs into HNix

`HNix` was taken and with maximal preservation of Git history, Git history was
passed though pass filter and that in how `haskell-with-nix` was formed.

Further work on CI and build configs happened there.

Due to `HNix` being a mothership project and since `haskell-with-nix`
preservation of original history - it is possible to merge progress from it
back.

Now, when `haskell-with-nixpkgs` main setup stabilized - it is ready to be
merged back, meanwile `HNix` needed to dodge Nixpkgs downstream quirks which
blocked the development processes and new CI was needed.

Merge went pretty ideal, for merge `haskell-with-nixpkgs` only relevant history
was also filtered-out and also dropped 3 commits that got into the passed files.

Custom merge commit also was required.

But overall, the process of merge went unexpectedly great.

---

After the merge - also needed the the change of stub names used in
`haskell-with-nixpkgs` to `HNix`-specific names. So that directly follows this
merge inside PR branch.

Anton-Latukha committed:
CI: GitHub: Cabal-Linux: provide libsodium

Anton-Latukha committed:
default.nix: upd neat-interpolation override

Anton-Latukha committed:
build.sh: clean-up

rm redundant settings, since default.nix is the main API propagation point

Anton-Latukha committed:
default.nix: set proper cabalName

Anton-Latukha committed:
CI: switch back to HNix Cachix account

M  .github/workflows/Nixpkgs-GHCJS.yml
M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml
M  .github/workflows/Nixpkgs-macOS.yml
M  .travis.yml

Anton-Latukha committed:
.travis.yml: switch back to HNix Cachix account secret

Anton-Latukha committed:
default.nix: upd GHC to 8.8.4

Anton-Latukha committed:
CI: update GHC 8.8.(3->4)

Also dropping minor numbers from build names for bravity of tracking Nixpkgs

M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml
M  .travis.yml

Anton-Latukha committed:
CI: rm Nixpkgs GHC 8.4.4 builds

8.4.4 support is removed from Nixpkgs

M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .travis.yml

Anton-Latukha committed:
CI: GitHub: add submodule recursion to checkouts

M  .github/workflows/Cabal-Linux.yml
M  .github/workflows/Nixpkgs-GHCJS.yml
M  .github/workflows/Nixpkgs-Linux-additional.yml
M  .github/workflows/Nixpkgs-Linux-main.yml
M  .github/workflows/Nixpkgs-macOS.yml

Anton-Latukha committed:
CI: Cabal-Linux: add GHC 8.4 build

Since Nixpkgs removed 8.4, adding it to Hackage builds.

Anton-Latukha committed:
CI: Cabal-Linux: enable tests in configuration

M  .github/workflows/Cabal-Linux.yml

Anton-Latukha committed:
CI: Cabal-Linux: clean-up

M  .github/workflows/Cabal-Linux.yml

Anton-Latukha committed:
CI: Cabal-Linux: add Nix installation, instantiate still required for tests

M  .github/workflows/Cabal-Linux.yml

sjakobi commented on default.nix:
```diff
@@ -197,21 +198,22 @@ let
       # require n-i >= 0.4. dontCheck helps us avoid conflicts with
       # neat-interpolation's test dependencies.
-      neat-interpolation = pkgs.haskell.lib.dontCheck super.neat-interpolation_0_5_1;
+      neat-interpolation = pkgs.haskell.lib.dontCheck super.neat-interpolation_0_5_1_1;
```
Mustn't we assume that `neat-interpolation_0_5_1_1` will soon be gone too, when the next version of `neat-interpolation` is released or the latest version becomes the default?

sjakobi replied:
(That's why I had used `callPackage` in https://github.com/haskell-nix/hnix/pull/680)

Anton-Latukha replied:
If someone knows how we can circumvent this Nixpkgs flaw and always find import from the Nixpkgs the last release version of the package - I would be glad to know also.

Anton-Latukha replied:
The argument is to prefer Nixpkgs packages over custom overrides, because we test Nixpkgs builds, and if we override everything - that is not Nixpkgs test, that is custom Nix config test that has almost nothing with Nixpkgs.

Looking from the other side on the same argument:
Lets assume the contrary. Preferring override of the package to what is available in the store essentially means to prefer avoiding the global store for what it was created and prefer keeping custom store, then the question is why we even test for Nixpkgs then if we had chosen to avoid it.

So if we stick to Nixpkgs - we should try to follow it and be in touch in the processes in it. We may even have it as optional, but CI builds should track the state there.

Anton-Latukha replied:
Well, on the other side - I also do not like that the Nixpkgs quirks constantly break the CI pass through process.

As I said - lets make Nixpkgs optional. - That seems the most effective solution to the problem.

And after that we may think to compose some middle ground, we can have a build option with frozen Nixpkgs overrides, so if devs/users can not build with current Nixpkgs - README can direct them to use frozen version. Of course, we can just direct or expect them switch to other GHC sets if possible - that solves a part of that probabilistic issue, and of course we would need to maintain the frozen overrides if we provide them.

So currently making more Nixpkgs build optional seems the most effective solution.

sjakobi replied:
If we can make the Nixpkgs build optional, we might as well get rid of it.

I thought the basic problem was simply that the Nixpkgs revision is unpinned, and therefore each build might use a different and often incompatible configuration. I've never understood why we don't simply pin the revision and update it every few months or as needed.

Anton-Latukha replied:
But the meaning of Nixpkgs build is Nixpkgs build tracking. Otherwise - we can just `cabal2nix` into custom Nix file and forget about Nixpkgs at all.

The argument that optional build can be reduced, it is analog to saying: since monitoring is optional - we can remove monitoring, since documentation is optional for operations - we can remove documentation.

Anton-Latukha replied:
The reason for Nixpkgs to be optional - is that since it is brittle - we would see the changes that needs to be made, but CI would pass PRs at the same time, and we can see that Cabal & Hackage & testing suites work and that error relates to downstream Nipkgs issue and not to the PR code, and so would remind us to fix it.

The idea is so we shift from sequence requirement `(code merge request with CI block due to Nixpkgs -> wait -> develop & merge Nixpkgs fix PR -> wait -> rebase CI tests & code merge -> HNix)` develop and fix Nixpkgs in separate simple processes and in parrallel `(code merge request -> HNix),  (PR on Nixpkgs fix -> HNix)` since that can be done in parallel.

Anton-Latukha replied:
As I said, using of Nix for Nixpkgs testing is essentially indeed optional, it is HNix being a good citizen for Nix ecosystem. But removing the tests - is taking a revolver and becoming Nixpkgs Wild West outlaw, that feeds on Nix ecosystem but has no regard to its laws and we would be like:
```
"

- (Nix community)  You broke everything. it is broken for 3 month now.

- (HNix)  HNix does not work and breaks stuff in Nixpkgs? - honestly, we do not care to monitor it, we removed test builds for it, we do not care about it to such degree.

- (Nix community)  Wait what is the agenda of the project?

- (HNix)  To support and evaluate Nix ecosystem, most importantly Nixpkgs.

- (Nix community)   ... wait, the agenda of the project is to work with and within Nix ecosystem - right? And you chosen to not care about Nixpkgs?.. Isn't that a direct collision that you disregard the purpose of the project and the community it was did for?

- (HNix)  ... Look. Here is the revolver.

"
```

If we decide to stop even monitoring Nixpkgs, we as well can just turn around and leave the project now.

Anton-Latukha replied:
Regarding why we do not pin the Nixpkgs - the same question is "Why we do not freeze the Cabal configuration for Hackage."

Because there is no sense - if we pin Nixpkgs for our builds, then (very quickly) - the HNix would not work in new Nixpkgs, and then what is the sense in pinning Nixpkgs if all community constantly rolls on Nixpkgs and HNix does not work for them, so when we pin - we slash the community and the growth speed of it.

HNix can pin Nixpkgs to achieve interpreting Nixpgs successfully - then we internally stabilize the state so we can develop for it, but then we do not track interpretation of current Nixpkgs and would need to adjust to it afterward constantly anyway, since that is the main aim of the project.

1. So pinning for Nixpkgs to build HNix - results into us not supporting and not advertising project to Nixpkgs community which is where main community located.

2. And pinning Nixpkgs for interpretation in HNix - means project most probably does not work for current Nixpkgs and we, nor our users can not open Reports in Nixpkgs upstream, since HNix does not use current Nixpkgs but some old frozen one, and the first question everyone asks on debugging is "What version you are using, please check for the latest release and report only for the most recent release.", and since we are who pinned Nixpkgs internally - we are responsible for solving that and tracking latest Nixpkgs release, and we anyway would need to follow Nixpkgs always, since it is a technical target of the project.

sjakobi replied:
OK. I'm somewhat skeptical about the need to build with the latest Nixpkgs all the time.

My current understanding is that the Haskell packages in Nixpkgs are basically just downstream from Stackage. So it seems to me that we can simply try to keep Stackage happy, and let the Nixpkgs maintainers worry about the rest.

If you still want this monitoring, that's fine by me, but please ensure that it doesn't affect CI results and that new contributors can get a clean build with the Nix setup at all times.

Anton-Latukha replied:
How Nixpkgs can be downstream from the Stackage, if Nixpkgs autoimports downstream from the Hackage.

For example, HNix is on Hacakge, and not in Stackage, but in Nixpkgs, how it gets into Nixpkgs after we release it on Hackage, in the night it already appears in Nixpkgs.

`stack2nix` was working half way half a time, and it is recenly deprecated by creators, `stackage2nix` also abandoned project.

The `peti` build projects `hackage2nix` `cabal2nix`, Nixpkgs machinery uses them, process tied on regeneration of `configuration-hackage2nix.yaml` that is the entry point for auto imported Haskell Nixpkgs project descriptions. He has presentations on it, and indeed every 3:30 in the night autoimported Haskell packages in Nixpkgs sync to Hackage, that is why HNix updates in there and we see if it builds or there is breakage.

---

The thing is. Nix is always pain. FP OS management for many years ahead would be a pain, it would be pain until FP OS ecosystems mature (and it would be longer process then Haskell maturity, FP OS management is much harder ad-hock question and volume of particulars then the design of the programming language), it would be a pain until IT would not check to comply with it, I mean that their projects build in Nix or Guix. All hidden under the carpet troubles of IT would and ignorance to collaborate of its people, while hidden in classical approaches - would lay explicit marks on the newborn sick FP OS management child.

And as I stated, FP IT management is a much more complex topit - too much projects, project options, OS management (bootloaders, system level (all of systemd), all filesystems), build systems, languages, own language, own interpreter/compiler, own store, hardware particulars - there are too many questions and too many work to maintain all of it while FP OS management is so demanding for its quality - it would be in the process of maturing for a long time. There was a lot of scientific work for FP, but almost none for how to design FP IT management language and how to do FP IT management of all things.

It is and would be a constant PITA for a very long time, years and years. What I hope is that with migration on Haskell things would become bearable and velocity would have bigger acceleration, so much more exponentially rapid process would happen.

sjakobi replied:
> For example, HNix is on Hacakge, and not in Stackage, but in Nixpkgs, how it gets into Nixpkgs after we release it on Hackage, in the night it already appears in Nixpkgs.

Good question. I don't know. See https://github.com/haskell-nix/hnix/issues/676#issue-656518706 though, specifically

> The problem is, that we get an older neat-interpolation ==0.3.2.6 from stack.

Anton-Latukha replied:
That means that I need to learn more about Haskell Nix stack.

There is indeed Stack hooks.

Maybe part of the answer is that Stackage essentially is a Hackage, at least in terms of API, so same Hackage tools can download from both places.

Or most probably it is even simpler. Tool to get the consistent package set - receives a last LTS package version numbers from Stackage, and integrates those packages directly from Stackage or Hackage.

The full chain of `Hackage -> ... -> Nixpkgs install` is something I personally never thoroughly went through.

And actually, that is what I want to do, in previous talks I was talking to have a real Nixpkgs integration test, where the chain of integration would be a great representation of Hackage -> Nixpkgs integration build and test process, it would revolve around `hackage2nix`, `cabal2nix` and regeneration of `configuration-hackage2nix.yml`. So the integration test would closely replicate the process, that would sum up this our conversation.

This is my next agenda regarding the CI theme.

sjakobi commented on default.nix:
```diff
@@ -1,6 +1,13 @@
+#  2020-07-05: By default using default GHC for Nixpkgs, see https://search.nixos.org/packages?query=ghc&from=0&size=500&channel=unstable for current version (currently ghc883 == GHC 8.8.3)
+  compiler    ? "ghc884"
+
+# Deafult.nix is a unit package abstraciton that allows to abstract over packages even in monorepos:
+# Example: pass --arg cabalName --arg packageRoot "./subprojectDir", or map default.nix over a list of tiples for subprojects.
+# cabalName is package resulting name: by default and on error resolves in haskellPackages.developPackage to project root directory name by default, but outside the haskellPackages.developPackage as you see below packageRoot can be different
+, cabalName ? "hnix"
```
I don't understand the motivation for having this parameter. Could you give an example where it would be useful?

Anton-Latukha replied:
It is a ground work towards the support of monorepo with several projects.

Since `develoPackage` should include `name` of the Cabal file, and different projects have different Cabal names, passing through the `cabalName` and `projectRoot` - allows to traverse over the projects in monorepo. `hnix-store` is example of such Nix infrastructure - it has one `default.nix` for two projects.

Since HNix plans to move to monorepo at some near point - I preserved this ground work, there is also CI builds which matrix traverse over project roots `[ ./ ]`, aka singleton currently.

sjakobi replied:
> Since HNix plans to move to monorepo at some near point

Who is planning this? Is there an issue where this plan is discussed? If not, we should have one.

Anton-Latukha replied:
It is `"Split parser & types out into different packages"` #328, there the mororepo approach was discussed much, several people from their experience voted up for it, John also voted up that he ready moreover then monorepo - even completely split the lib from executable, what is to do - is of course to what is most effective, what most effective should be planned.

The backtracking `haskell-with-nixpkgs` monorepo support back for merge into HNix - overcomplicated the merge process. The monorepo preparation here is minimal, it is several lines of code in `default.nix` and in GitHub workflows, and it is explained in the comments.

The monorepo code is easy and minimal, and to remove monorepo preparation is several easy commits, so I left it, since it is done, documented, awaited in this project, and always can be easily removed.

I've looked into `hnix-store` and thought if we would go monorepo - most probably it would be how it is done in `hnix-store`.

sjakobi replied:
Thanks for the link. I think it would be good to have a separate issue for the mono-repo discussion. It's currently rather hidden in a very long thread.

sjakobi commented on .travis.yml:
```diff
@@ -21,8 +21,6 @@ git:
 
 env:
```
Why do we still have the Travis config? Isn't GitHub CI sufficient?

Anton-Latukha replied:
Yep, still need it.

For a Nixpkgs cache loophole, explained in: https://github.com/haskell-nix/hnix/issues/682#issuecomment-667540898

Travis currently builds and pushes Cachix Nixpkgs cache that GitHub CI uses, while secret is not provided.

sjakobi replied:
I'm not thrilled to babysit _two_ CI systems! Please get this fixed soon.

Anton-Latukha replied:
If someone should not be thrilled to babysit _two_ CI systems - is I. Since I did major work working on them and maintaining them and obvious that I am who would maintain them in the nearest future.

And would say even more. Since it is free tiers and our CIs in the future would have different paradigms (classical and FP, aka Hydra or Hercules or Sorkis CI) -  more CI systems - may be beneficial. Some CI would have good Windows testing, some would provide true FP Nix CI with all its abilities, some other services would provide supplementary functions.

I feel vibes that you may need to rest and relax. When we feel something or many things as important - often we would need to go on a walk and think about what is really important in life. I am the one that need to worry about CIs, not you. I am calm and reserved about all this, I just try to understand your argumentation, so I respond to you in the talk form, so we write back and forth to get understanding.

I gave you a strong argument why, Travis currently provides the cache for GitHub CI, until John puts the a secret key that only he holds and should hold. So currently there is Travis to produce the Nixpkgs cache that gets shared over Cachix so the GitHub CI does not rebuilds the Nixpkgs stack for 3 hours at every commit someone does in PR. It can not be fixed sooner then John appears and provides the secret key for us, until then all our Nixpkgs builds tied to Travis CI.

---

For example, see the response of Gabriel: https://github.com/dhall-lang/dhall-haskell/issues/1678#issuecomment-592960057

Sometimes 3-4 CIs are the best stable setup that Free Software Haskell project can do currently.

Anton-Latukha replied:
Since Travis CI is supplementary at this point - I would make it optional and its check a fast skip, until we have normal cache process, so we would need to wait only on GitHub CI to pass, and Travis would always be green while would build cache at the moment.

sjakobi replied:
Sorry, @Anton-Latukha. It's hard for me stay relaxed about this when I receive an email from Travis CI informing me about the broken `master` branch every single day. Situations like the one in https://github.com/haskell-nix/hnix/pull/680 are also very frustrating to me because the CI failure appears to be completely unrelated to my work and there is no information on how the build could be fixed.

I very much hope that you will put an end to this misery.

sjakobi replied:
> I gave you a strong argument why, Travis currently provides the cache for GitHub CI, until John puts the a secret key that only he holds and should hold. So currently there is Travis to produce the Nixpkgs cache that gets shared over Cachix so the GitHub CI does not rebuilds the Nixpkgs stack for 3 hours at every commit someone does in PR. It can not be fixed sooner then John appears and provides the secret key for us, until then all our Nixpkgs builds tied to Travis CI.

Ping @jwiegley.

Anton-Latukha replied:
Well, no hurry, I already told him a couple of times before, pinged him a couple of times, and created appropriate Report #682, it is not urgent. I've also could of course directly require this action in advance.

I've just made all Travis builds "optional" and enabled "fast skip" of them, and added a "bypass build" that returns "exit" as soon as instance boots up, so Travis always green and passes in 40 seconds for us, and builds and provides caches in the background.

sjakobi commented on .github/workflows/Cabal-Linux.yml:
```diff
@@ -0,0 +1,67 @@
+      - run: cabal v2-test
+
+
+  build20:
```
What does `build20` mean? If it's just a name, how about referencing the main property/properties of this job?, e.g. `ghc804`.

Anton-Latukha replied:
Well, there is `name` field for it. ID of the builds and `name` are mandatory, and essentially fields that are duplicates, but required by the GitHub CI design.

GitHub CI has number of design choises I am complettely puzzled about. Like they not follow basic YAML RFC and so metadata build IDs strings are tightly restricted set of symbols, and inside there is the separate `name` field which follows YAML RFC, if they respected YAML RFC in IDs - that would've been one field.

There is `name` field, there is already manual tracking of renaming needed, since there already a `name` - metadata IDs can not bear any useful information only try to duplicate the `name` all the time, and because one has more important stuff then to duplicate sync same information translating it into ID required restricted form - made a choise for the `name` to shows the real human readable name and required for machines ID field is just old-style unstractured-programming type numbering that allows to put 9 builds in between them, since IDs does not really metter.

sjakobi replied:
Thanks for the explanation!

sjakobi reviewed (commented):
Wow! Looks like a massive amount of work! Will this help make CI more reliable, i.e. more robust against Nixpkgs changes and build caching issues?

Also, why is the GitHub CI config spread over 5 files? I assume that there must be a lot of duplication that might be tricky to keep in sync.

Anton-Latukha committed:
CI: Cabal-Linux: make builds pass mandatory

Anton-Latukha committed:
CI: Nixpkgs-Linux-main: make all builds optional

Due to Nixpkgs brittleness - it often blocks the development process, when the
PRs do not pass due to issues of downstream Nixpkgs.

Moreover that HNix builds in Nixpkgs with Nix, so the point of this builds is
just to check that HNix builds with Nixpkgs under Nix.

When Nixpkgs build is mandatory - PRs get blocked by it and PRs await on fixing
Nixpkgs.

Making Nixpkgs builds optional allows parallelization of code development and
Nixpkgs fixing.

More in discussion: https://github.com/haskell-nix/hnix/pull/681#discussion_r463983376

Anton-Latukha committed:
CI: Travis: enable fast skip

See: https://github.com/haskell-nix/hnix/pull/681#discussion_r463996365

M  .travis.yml

Anton-Latukha committed:
CI: Travis: add "bypass" to keep CI green, send cache gen to background

M  .travis.yml

Anton-Latukha committed:
CI: Nixpkgs-Linux-main: make all builds optional cache builders

Due to Nixpkgs brittleness - it often blocks the development process, when the
PRs do not pass due to issues of downstream Nixpkgs.

Moreover that HNix builds in Nixpkgs with Nix, so the point of this builds is
just to check that HNix builds with Nixpkgs under Nix.

When Nixpkgs build is mandatory - PRs get blocked by it and PRs await on fixing
Nixpkgs.

Making Nixpkgs builds optional allows parallelization of code development and
Nixpkgs fixing.

More in discussion: https://github.com/haskell-nix/hnix/pull/681#discussion_r463983376

CI: Travis: fx optional builds

CI: Nixpkgs-Linux-main: make all builds optional

Due to Nixpkgs brittleness - it often blocks the development process, when the
PRs do not pass due to issues of downstream Nixpkgs.

Moreover that HNix builds in Nixpkgs with Nix, so the point of this builds is
just to check that HNix builds with Nixpkgs under Nix.

When Nixpkgs build is mandatory - PRs get blocked by it and PRs await on fixing
Nixpkgs.

Making Nixpkgs builds optional allows parallelization of code development and
Nixpkgs fixing.

More in discussion: https://github.com/haskell-nix/hnix/pull/681#discussion_r463983376

CI: Travis: fx optional builds

Anton-Latukha committed:
CI: Nixpkgs-Linux-main: proper executable name to shell complete

Anton-Latukha committed:
CI: Nixpkgs-Linux-main: Quality build: add shell completion test

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
CI: Nixpkgs-Linux-main: default build: add strict build

M  .github/workflows/Nixpkgs-Linux-main.yml

Anton-Latukha committed:
CI: Nixpkgs-GHCJS: print log tail

Anton-Latukha committed:
CI: Nixpkgs-GHCJS: turn-off build which is not persued

sjakobi reviewed (approved):
LGTM based on my impression that this addresses some of our current CI woes.

If there are any issues we'll probably detect them while using the CI in anger.

Adding hundreds of commits to `master` for this CI overhaul seems a little disproportionate, so maybe this PR should be squashed.

Thanks for all the work, @Anton-Latukha, and thanks for addressing my concerns with the current CI setup! :+1:
